File: apps/api/src/notifications/notifications.service.ts
import { Injectable, Logger } from "@nestjs/common";
import { PrismaService } from "../common/prisma/prisma.service";
import { EmailClient } from "./email.client";
import { SmsClient } from "./sms.client";
import { WhatsAppClient } from "./whatsapp.client";

export type NotificationType =
  | "booking_confirmed"
  | "payment_receipt"
  | "agreement_signing_link"
  | "move_in_reminder"
  | "instalment_reminder"
  | "cancellation"
  | "refund_processed"
  | "support_status";

interface NotifyOptions {
  userId: string;
  type: NotificationType;
  subject: string;
  body: string;
  payload?: Record<string, unknown>;
}

@Injectable()
export class NotificationsService {
  private readonly logger = new Logger(NotificationsService.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly emailClient: EmailClient,
    private readonly smsClient: SmsClient,
    private readonly whatsAppClient: WhatsAppClient
  ) {}

  /**
   * Dispatches a notification across every channel the user has opted into
   * (Part D: `notification_preferences`). Each channel is attempted
   * independently — one channel's provider being unconfigured or failing
   * never blocks the others, and every attempt (sent or not) is recorded as
   * a `Notification` row for support/audit purposes.
   */
  async notify(options: NotifyOptions): Promise<void> {
    const user = await this.prisma.user.findUnique({ where: { id: options.userId } });
    if (!user) {
      this.logger.warn(`notify() called for a nonexistent user ${options.userId}; skipping.`);
      return;
    }

    const preference = await this.prisma.notificationPreference.findUnique({
      where: { userId: options.userId },
    });
    // Defaults match the schema's own column defaults if no row exists yet:
    // email/SMS opt-in true, WhatsApp opt-in false.
    const emailOptIn = preference?.emailOptIn ?? true;
    const smsOptIn = preference?.smsOptIn ?? true;
    const whatsappOptIn = preference?.whatsappOptIn ?? false;

    const attempts: Array<{ channel: string; enabled: boolean; send: () => Promise<void> }> = [
      {
        channel: "email",
        enabled: emailOptIn && Boolean(user.email),
        send: () => this.emailClient.send(user.email!, options.subject, options.body),
      },
      {
        channel: "sms",
        enabled: smsOptIn && Boolean(user.phone),
        send: () => this.smsClient.send(user.phone!, options.body),
      },
      {
        channel: "whatsapp",
        enabled: whatsappOptIn && Boolean(user.phone),
        send: () => this.whatsAppClient.send(user.phone!, options.body),
      },
    ];

    for (const attempt of attempts) {
      if (!attempt.enabled) continue;

      const notification = await this.prisma.notification.create({
        data: {
          userId: options.userId,
          channel: attempt.channel,
          type: options.type,
          payloadJson: options.payload as any,
        },
      });

      try {
        await attempt.send();
        await this.prisma.notification.update({
          where: { id: notification.id },
          data: { sentAt: new Date() },
        });
      } catch (err) {
        // A provider being unconfigured/down never throws out of notify()
        // itself — a notification failure must never break the booking,
        // payment, or agreement flow that triggered it.
        this.logger.error(
          `Failed to send ${attempt.channel} notification (${options.type}) to user ${options.userId}: ${
            err instanceof Error ? err.message : String(err)
          }`
        );
      }
    }
  }
}

File: apps/api-py/app/services/notifications_service.py
"""Direct port of notifications/notifications.service.ts's notify() method."""

import logging
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Notification, NotificationPreference, User
from app.services import notification_clients

logger = logging.getLogger(__name__)


async def notify(
    db: AsyncSession,
    user_id: str,
    notification_type: str,
    subject: str,
    body: str,
    payload: dict | None = None,
) -> None:
    """Dispatches a notification across every channel the user has opted
    into. Each channel is attempted independently — one channel's provider
    being unconfigured or failing never blocks the others, and every
    attempt (sent or not) is recorded as a Notification row for
    support/audit purposes."""
    user = (await db.execute(select(User).where(User.id == user_id))).scalar_one_or_none()
    if user is None:
        logger.warning("notify() called for a nonexistent user %s; skipping.", user_id)
        return

    preference = (
        await db.execute(select(NotificationPreference).where(NotificationPreference.user_id == user_id))
    ).scalar_one_or_none()
    email_opt_in = preference.email_opt_in if preference else True
    sms_opt_in = preference.sms_opt_in if preference else True
    whatsapp_opt_in = preference.whatsapp_opt_in if preference else False

    attempts = [
        {"channel": "email", "enabled": email_opt_in and bool(user.email), "send": lambda: notification_clients.send_email(user.email, subject, body)},
        {"channel": "sms", "enabled": sms_opt_in and bool(user.phone), "send": lambda: notification_clients.send_sms(user.phone, body)},
        {"channel": "whatsapp", "enabled": whatsapp_opt_in and bool(user.phone), "send": lambda: notification_clients.send_whatsapp(user.phone, body)},
    ]

    for attempt in attempts:
        if not attempt["enabled"]:
            continue

        notification = Notification(user_id=user_id, channel=attempt["channel"], type=notification_type, payload_json=payload)
        db.add(notification)
        await db.commit()
        await db.refresh(notification)

        try:
            await attempt["send"]()
            notification.sent_at = datetime.now(timezone.utc)
            await db.commit()
        except Exception as err:
            # A provider being unconfigured/down never throws out of
            # notify() itself — a notification failure must never break
            # the booking/payment/agreement flow that triggered it.
            logger.error(
                "Failed to send %s notification (%s) to user %s: %s",
                attempt["channel"], notification_type, user_id, err,
            )
